core/shared/src/main/scala/zio/internal/FiberRuntime.scala

```diff
@@ -1063,23 +1063,26 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
             case flatmap: FlatMap[Any, Any, Any, Any] =>
               updateLastTrace(flatmap.trace)
 
-              stackIndex = pushStackFrame(flatmap, stackIndex)
+              val first = flatmap.first
 
-              val result = runLoop(flatmap.first, stackIndex, stackIndex, currentDepth + 1, ops)
-              ops += 1
-
-              if (null eq result)
-                return null
+              if (first eq ZIO.unit) cur = flatmap.successK(())
+              else if (first.isInstanceOf[Success[Any]]) cur = flatmap.successK(first.asInstanceOf[Success[Any]].value)
+              else if (first.isInstanceOf[Failure[Any]]) cur = first
               else {
-                stackIndex -= 1
-                popStackFrame(stackIndex)
+                stackIndex = pushStackFrame(flatmap, stackIndex)
 
-                result match {
-                  case s: Success[Any] =>
-                    cur = flatmap.successK(s.value)
+                val result = runLoop(flatmap.first, stackIndex, stackIndex, currentDepth + 1, ops)
+                ops += 1
 
-                  case failure =>
-                    cur = failure
+                if (null eq result) return null
+                else {
```

Comment:
While we're at it, let's remove this `else` statement. It doesn't make sense since `return` will exit the method anyways

Reply:
Done